### ISSUE TITLE
API design: imlement Future for Vec<Future>?

### ISSUE THREAD

**Issue by yoshuawuyts** [async-rs/async-std#424]
We currently don't have an equivalent of `future::join_all`. I was thinking: what if we had an implementation that did the same by implementing Future on a Vec of Futures:

```rust
impl<T> Future for Vec<Future<Output = T>> {
    type Output = Vec<T>;
}

impl<T, E> Future for Vec<Future<Output = Result<T, E>> {
    type Output = Result<Vec<T>, E>;
}
```

Not sure if orphan/coherence rules allow this tho. But this might be a nice shorthand. Also interesting perhaps to explore if we can enable this for arrays too. And whether we could use `IntoFuture` and `IntoIteraror` in the bounds rather than `Future` and `Vec` (probably yes on the first account, no on the second).

**Comment by stjepang** [comment]
That's a very interesting idea. I think such a blanket impl could only live in the standard library. Not sure if there are any coherence issues, but it would probably work!

Is `future::join_all(v)` essentially equivalent to this?
```rust
v.into_stream().map(|f| f).collect()`
```

**Comment by yoshuawuyts** [comment]
@stjepang not quite, as in `join_all!` the futures would resolve concurrently (though being ordered!), whereas in `Stream::map` they resolve sequentially.

**Comment by stjepang** [comment]
What if we made it possible to resolve concurrently by doing the following:

```rust
v.into_stream().par().map(|f| f).collect()
```

**Comment by skade** [comment]
IMHO, the result of this should be `Box<[T]>` and the function implemented on `&[T]`. I don't think the input and output of this should be manipulated or grown by the user.

I would also like to see a parallel API like @stjepang suggests.
